stellargraph/core/element_data.py

```diff
@@ -335,29 +381,62 @@ def __init__(self, ids, sources, targets, weights, type_starts):
         self._empty_ilocs = np.array([], dtype=np.uint8)
 
     def _init_directed_adj_lists(self):
-        # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            in_dict.setdefault(tgt, []).append(i)
-            out_dict.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_in_dict = _numpyise(in_dict, dtype=dtype)
-        self._edges_out_dict = _numpyise(out_dict, dtype=dtype)
+        # record the edge ilocs of incoming and outgoing edges
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        def _to_dir_adj_list(arr):
+            neigh_counts = np.bincount(arr, minlength=self.number_of_nodes)
+            neigh_counts = neigh_counts.astype(self._id_index.dtype, copy=False)
+            splits = np.cumsum(neigh_counts, dtype=self._id_index.dtype)
+            flat = np.argsort(arr).astype(self._id_index.dtype, copy=False)
+            return FlatAdjacencyList(flat, splits)
+
+        self._edges_in_dict = _to_dir_adj_list(self.targets)
+        self._edges_out_dict = _to_dir_adj_list(self.sources)
 
     def _init_undirected_adj_lists(self):
-        # record the edge ilocs of incoming, outgoing and both-direction edges
-        undirected = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            undirected.setdefault(tgt, []).append(i)
-            if src != tgt:
-                undirected.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_dict = _numpyise(undirected, dtype=dtype)
+        # record the edge ilocs of both-direction edges
+        num_edges = len(self.targets)
+        if num_edges == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        # the dtype of the edge_ilocs
+        _dtype = np.min_scalar_type(2 * len(self.sources))
```

Comment:
Also, given it's so subtle, it'd be good to have a test that validates this explicitly, e.g. a circle graph with 200 nodes (and thus 200 edges). The undirected adjacency ilocs initially computed by the `argsort` need to have dtype uint16, but the main edge ilocs will have dtype uint8. If this function is implemented incorrectly, the argsort's `astype` call may cause truncation and break the connections. (To make sure the test is implemented correctly and testing the right thing, it'd be good to check it fails after removing the `2 *` here.)

Reply:
yeah good point, ive added this test in and checked that the connectivity gets missed up with the wrong dtype along with some comments